File: solutions/haskell/perfect-numbers/2/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification (..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify num =
    if num < 1
    then Nothing
    else classify' num

classify' :: Int -> Maybe Classification
classify' num
    | factorsSum num < num = Just Deficient
    | factorsSum num > num = Just Abundant
    | otherwise = Just Perfect

factorsSum :: Integral a => a -> a
factorsSum number = sum (factors number)

factors :: Integral a => a -> [a]
factors number =
    [divisor | divisor <- [1 .. (number `quot` 2)], number `rem` divisor == 0]

File: solutions/csharp/perfect-numbers/2/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0) throw new ArgumentOutOfRangeException();

        var aliquotSum = GetAliquotSum(number);

        if (aliquotSum < number) 
            return Classification.Deficient;
        else if (aliquotSum > number) 
            return Classification.Abundant;
        else 
            return Classification.Perfect;
    }

    private static int GetAliquotSum(int number) => Enumerable.Range(1, number/2).Where(factor => number % factor == 0).Sum();
}
